src/onnx_ir/passes/common/deduplicate_initializers.py

```diff
@@ -0,0 +1,69 @@
+# Copyright (c) ONNX Project Contributors
+# SPDX-License-Identifier: Apache-2.0
+"""Pass for removing duplicated initializer tensors from a graph."""
+
+from __future__ import annotations
+
+__all__ = [
+    "DeduplicateInitializersPass",
+]
+
+import hashlib
+
+import onnx_ir
+import onnx_ir.traversal
+
+
+class DeduplicateInitializersPass(onnx_ir.passes.InPlacePass):
+    """Remove duplicated initializer tensors from the graph.
+
+    This pass detects initializers with identical shape, dtype, and content,
+    and replaces all duplicate references with a canonical one.
+
+    For efficiency, it uses a hash of tensor bytes to group candidates,
+    then confirms exact match using the full byte content (to avoid collisions).
+    Subgraphs are handled via RecursiveGraphIterator.
+    """
+
+    def call(self, model: onnx_ir.Model) -> onnx_ir.passes.PassResult:
+        graph = model.graph
+        seen: dict[tuple[str, tuple[int, ...]], dict[int, list[tuple[str, bytes]]]] = {}
+
+        name_map = {}  # Duplicate name → canonical name
+
+        for initializer in list(graph.initializers.values()):
+            const_val = initializer.const_value
+            if const_val is None:
+                continue  # Skip if initializer has no constant value
+            dtype = const_val.dtype
+            shape = tuple(const_val.shape)
+            content = const_val.tobytes()
+            content_hash = hashlib.sha256(content).hexdigest()
+
+            key = (dtype, shape)
+            if key not in seen:
+                seen[key] = {}
+
+            group = seen[key]
+            if content_hash in group:
+                # Verify collision using full bytes
+                for existing_name, existing_bytes in group[content_hash]:
+                    if existing_bytes == content:
+                        name_map[initializer.name] = existing_name
+                        if initializer.name is not None:
+                            graph.initializers.pop(initializer.name)
+                        break  # only break when deduplication is successful
+                else:
+                    # no matching content found: append as a new entry
+                    group[content_hash].append((initializer.name, content))
```

Comment:
You may store the values instead in the hash table, and use const_val.tobytes() for comparison. This way the bytes do not stay in the memory or take up space?

```suggestion
                    group[content_hash].append(const_val)
```